execution-plane/arcon/src/streaming/source/kafka.rs

```diff
@@ -1,41 +1,281 @@
+            let mut timestamp: Option<u64> = None;
+            if let Some(ts) = m.timestamp().to_millis() {
+                timestamp = Some(ts as u64);
+                if ts as u64 > self.max_timestamp {
+                    self.max_timestamp = ts as u64;
+                }
+            }
+            if let Ok(data) = serde_json::from_str(&payload) {
+                self.output_event(data, timestamp);
+            } else {
+                error!(self.ctx.log(), "Unable to deserialize message:\nkey: '{:?}', payload: '{}', topic: {}, partition: {}, offset: {}, timestamp: {:?}",
+                      m.key(), payload, m.topic(), m.partition(), m.offset(), m.timestamp());
+            }
+            /*
+            if let Some(headers) = m.headers() {
+                for i in 0..headers.count() {
+                    let header = headers.get(i).unwrap();
+                    println!("  Header {:#?}: {:?}", header.0, header.1);
+                }
+            }
+            // todo (somewhere else) 
+            consumer.commit_message(&m, CommitMode::Async).unwrap(); auto-committing
+            */
+        }
+        if counter > 0 {
+            self.output_watermark();
+            // Store Offset manage epoch here.
+        }
+        // Schedule next batch
+        self.schedule_once(Duration::from_millis(1000), move |self_c, _|{self_c.receive()});
+    }
+}
 
-*/
-#[derive(ComponentDefinition)]
-pub struct KafkaSource<A: 'static + ArconType> {
-    ctx: ComponentContext<KafkaSource<A>>,
-    target_pointer: ActorRef,
-    checkpoint_interval: u64,
-    _output_format: A,
+impl<OUT> Provide<ControlPort> for KafkaSource<OUT>
+where
+    OUT: 'static + ArconType + DeserializeOwned,
+{
+    fn handle(&mut self, event: ControlEvent) -> () {
+        match event {
+            ControlEvent::Start => {
+                self.receive();
+            }
+            _ => {
+                error!(self.ctx.log(), "bad ControlEvent");
+            }
+        }
+    }
 }
 
-impl<A: ArconType> Provide<ControlPort> for KafkaSource<A> {
-    fn handle(&mut self, event: ControlEvent) -> () {}
+impl<OUT> Actor for KafkaSource<OUT>
+where
+    OUT: 'static + ArconType + DeserializeOwned,
+{
+    fn receive_local(&mut self, _sender: ActorRef, _msg: &Any) {
+    }
+    fn receive_message(&mut self, _sender: ActorPath, _ser_id: u64, _buf: &mut Buf) {
+    }
 }
-impl<A: ArconType> Actor for KafkaSource<A> {
-    fn receive_local(&mut self, _sender: ActorRef, msg: &Any) {}
 
-    fn receive_message(&mut self, _sender: ActorPath, ser_id: u64, buf: &mut Buf) {}
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::{thread, time};
+    // Stub for window-results
+    #[arcon]
+    struct Thing {
+        id: u32,
+        attribute: i32,
+        location: Point,
+    }
+    #[arcon]
+    struct Point {
+        x: f32,
+        y: f32,
+    }
+    // JSON Example: {"id":1, "attribute":-13,"location":{"x":0.14124,"y":5882.231}}
+    mod sink {
```

Comment:
Is this needed? is it possible to use the Debug sink instead?

Reply:
Yeah I noticed this as well while working on the next commit. Already working on changing all the tests to use the DebugSink 👍